Use enum classes for QgsProperty property keys

nyalldawson opened the pull request:
Should fix usage on Qt 6 builds

nyalldawson committed:
test masks

nyalldawson committed:
Use enum classes for QgsProperty property keys

Should fix usage on Qt 6 builds

rouault commented on src/3d/materials/qgsgoochmaterialsettings.cpp:
```diff
@@ -148,10 +148,10 @@ void QgsGoochMaterialSettings::addParametersToEffect( Qt3DRender::QEffect * ) co
-  const QColor warm = dataDefinedProperties().valueAsColor( Warm, expressionContext, mWarm );
-  const QColor cool = dataDefinedProperties().valueAsColor( Cool, expressionContext, mCool );
-  const QColor specular = dataDefinedProperties().valueAsColor( Specular, expressionContext, mSpecular );
+  const QColor diffuse = dataDefinedProperties().valueAsColor( static_cast< int >( QgsAbstractMaterialSettings::Property::Diffuse ), expressionContext, mDiffuse );
```
all those explicit casts to int are quite ugly. Isn't there some template based solution that could automatically do it.

like 
```
template<class T> QColor valueAsColor( const T& key, const QgsExpressionContext &context, const QColor &defaultColor = QColor(), bool *ok = nullptr ) const
{
    return valueAsColor(static_cast<int>(key), context, defaultColor, ok);
}
```

nyalldawson replied:
Oh yeah, that's MUCH cleaner! :100:

nyalldawson committed:
Use templates to avoid ugly static_cast

nyalldawson committed:
Fix doxygen warning

nyalldawson committed:
Promote another enum to enum class

nyalldawson committed:
Fix build

nyalldawson committed:
Add missing files

nyalldawson committed:
Enable more tests

nyalldawson committed:
Use correct sip method

And use INDENT-OFF annotations to avoid astyle messing up
formatting of enum with inline macros
